samples/scan/scan.cpp

```diff
@@ -57,104 +57,210 @@ struct identity<T, std::logical_and<T>> {
+              temp[i + off - 1] = u;
+              temp[i + off * 2 - 1] = Op{}(t, u);
+            }
+          }
+
+          // Synchronize again to observe results.
+          item.barrier(sycl::access::fence_space::local_space);
+
+          /* To return an inclusive rather than exclusive scan result, shift
+           * each element left by 1 when writing back into global memory. If
+           * we are the last work-item, also add on the final element. */
+          data[2 * gid] = temp[2 * lid + 1];
+
+          if (lid == wgroup_size - 1) {
+            data[2 * gid + 1] = Op{}(temp[2 * lid + 1], second_in);
+          } else {
+            data[2 * gid + 1] = temp[2 * lid + 2];
+          }
+        });
+  });
+
+  // At this point we have computed the inclusive scans of this many segments.
+  size_t n_segments = half_in_size / wgroup_size;
+
+  if (n_segments == 1) {
+    // If all of the data is in one segment, we're done.
+    return;
+  }
+  // Otherwise we have to propagate the scan results forward into later
+  // segments.
+
+  // Allocate space for one (last) element per segment.
+  sycl::buffer<T, 1> ends{sycl::range<1>(n_segments)};
+
+  // Store the elements in this space.
+  q.submit([&](sycl::handler& cgh) {
+    auto scans = in.template get_access<sycl::access::mode::read>(cgh);
+    auto elems =
+        ends.template get_access<sycl::access::mode::discard_write>(cgh);
+
+    cgh.parallel_for<kernel_name<T, Op, 1>>(
+        sycl::range<1>(n_segments), [=](sycl::item<1> item) {
+          auto id = item.get_linear_id();
+          // Offset into the last element of each segment.
+          elems[item] = scans[(id + 1) * 2 * wgroup_size - 1];
+        });
+  });
+
+  // Recursively scan the array of last elements.
+  par_scan<T, Op>(ends, q);
+
+  // Add the results of the scan to each segment.
+  q.submit([&](sycl::handler& cgh) {
+    auto ends_scan = ends.template get_access<sycl::access::mode::read>(cgh);
+    auto data = in.template get_access<sycl::access::mode::read_write>(cgh);
+
+    cgh.parallel_for<kernel_name<T, Op, 2>>(
+        // Work with one less work-group, since the first segment is correct.
+        sycl::nd_range<1>(half_in_size - wgroup_size, wgroup_size),
+        [=](sycl::nd_item<1> item) {
+          auto group = item.get_group_linear_id();
+
+          // Start with the second segment.
+          auto off_gid = item.get_global_linear_id() + wgroup_size;
+
+          /* Each work-group adds the corresponding number in the
+           * "last element scan" array to every element in the group's
+           * segment. */
+          data[off_gid * 2] = Op{}(data[off_gid * 2], ends_scan[group]);
+          data[off_gid * 2 + 1] = Op{}(data[off_gid * 2 + 1], ends_scan[group]);
+        });
+  });
+}
 
 /* Tests the scan with an addition operation, which is its most common use.
  * Returns 0 if successful, a nonzero value otherwise. */
 int test_sum(sycl::queue& q) {
+  constexpr size_t SIZE = 512;
```

Comment:
I think you don't need to capitalize this - it's a local var, not a macro, as much as it is constexpr.